<a href="https://colab.research.google.com/github/KumundzhievMaxim/AppliedDeepLearning/blob/main/EmotionRecognition/emotion_recognition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Downloading the Data {When UNI Link unreacheable}

In [ ]:
# Audio Song 1-24
!wget -O Audio_Song_Actors_01-24.zip https://zenodo.org/record/1188976/files/Audio_Song_Actors_01-24.zip?download=1 

--2020-12-15 07:30:35--  https://zenodo.org/record/1188976/files/Audio_Song_Actors_01-24.zip?download=1
Resolving zenodo.org (zenodo.org)... 137.138.76.77
Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 225505317 (215M) [application/octet-stream]
Saving to: ‘Audio_Song_Actors_01-24.zip’

Audio_Song_Actors_0 100%[===================>] 215.06M  7.43MB/s    in 31s     

2020-12-15 07:31:07 (7.04 MB/s) - ‘Audio_Song_Actors_01-24.zip’ saved [225505317/225505317]



In [ ]:
# Audio Speech
!wget -O Audio_Speech_Actors_01-24.zip https://zenodo.org/record/1188976/files/Audio_Speech_Actors_01-24.zip?download=1

--2020-12-15 07:31:57--  https://zenodo.org/record/1188976/files/Audio_Speech_Actors_01-24.zip?download=1
Resolving zenodo.org (zenodo.org)... 137.138.76.77
Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 208468073 (199M) [application/octet-stream]
Saving to: ‘Audio_Speech_Actors_01-24.zip’

Audio_Speech_Actors 100%[===================>] 198.81M  8.33MB/s    in 27s     

2020-12-15 07:32:25 (7.41 MB/s) - ‘Audio_Speech_Actors_01-24.zip’ saved [208468073/208468073]



In [ ]:
# Unzip Audio Song 1-24 (./song)
!mkdir ./song && unzip /content/Audio_Song_Actors_01-24.zip -d ./song && rm -r /content/Audio_Song_Actors_01-24.zip

mkdir: cannot create directory ‘./song’: File exists


In [ ]:
# Unzip Audio Speech 1-24 (./speech)
!mkdir ./speech && unzip -q /content/Audio_Speech_Actors_01-24.zip -d ./speech  && rm -r /content/Audio_Speech_Actors_01-24.zip

In [ ]:
# Downalod Video Song Actor 1-9
import os

for video_index in range(1, 10):
  down_cmd = f'wget -O Video_Song_Actor_0{video_index}.zip https://zenodo.org/record/1188976/files/Video_Song_Actor_0{video_index}.zip?download=1'
  dir_cmd = f'mkdir ./video/Actor_0{video_index}'
  unzip_cmd = f'unzip -q Video_Song_Actor_0{video_index}.zip -d ./video/Actor_0{video_index}/'
  rm_cmd = f'rm -r Video_Song_Actor_0{video_index}.zip'

  print(f'Downloading Video_{video_index} zip')
  os.system(down_cmd)
  
  print('Created folder for particular Video zip')
  os.system(dir_cmd)
  
  print('Unzipping particular Video')
  os.system(unzip_cmd)

  print('Deleting particular Video zip')
  os.system(rm_cmd)


Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip
Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip
Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip
Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip


# ENTRY POINT

## Data Description
The RAVDESS contains **7356** files. 
Each file was rated **10 times** on emotional validity, intensity, and genuineness

[Video example](https://www.youtube.com/watch?v=Y7OQoNEu3dY)

The database contains 24 professional actors (12 female, 12 male), vocalizing two lexically-matched statements in a neutral North American accent. Speech includes calm, happy, sad, angry, fearful, surprise, and disgust expressions.


All conditions are available in three modality formats: Audio-only (16bit, 48kHz .wav), Audio-Video (720p H.264, AAC 48kHz, .mp4), and Video-only (no sound).  <b>Note, there are no song files for Actor_18.</b>


### Names convetions and details
#### Audio-only files {Audio_Speech_Actors}
Audio-only files of all actors (01-24) are available as two separate zip files (~200 MB each):

Speech file (Audio_Speech_Actors_01-24.zip, 215 MB) contains 1440 files: 60 trials per actor x 24 actors = 1440. 

#### Audio-Visual and Video-only files files {Video_Speech_Actor}
Speech files (Video_Speech_Actor_01.zip to Video_Speech_Actor_24.zip) collectively contains 2880 files: 60 trials per actor x 2 modalities (AV, VO) x 24 actors = 2880.


### Download Data when UNI Link reachable

In [ ]:
!wget -O ravdess.zip http://nipg1.inf.elte.hu:8765/ravdess.zip 

--2020-12-15 14:52:48--  http://nipg1.inf.elte.hu:8765/ravdess.zip
Resolving nipg1.inf.elte.hu (nipg1.inf.elte.hu)... 157.181.160.161
Connecting to nipg1.inf.elte.hu (nipg1.inf.elte.hu)|157.181.160.161|:8765... connected.
HTTP request sent, awaiting response... 200 OK
Length: 13565818088 (13G) [application/zip]
Saving to: ‘ravdess.zip’

ravdess.zip         100%[===================>]  12.63G  16.2MB/s    in 11m 44s 

2020-12-15 15:04:33 (18.4 MB/s) - ‘ravdess.zip’ saved [13565818088/13565818088]



In [3]:
!unzip -q ravdess.zip 

### HERE

In [81]:
!pip install librosa soundfile pyaudio

  ERROR: Failed building wheel for pyaudio
  Running setup.py clean for pyaudio
Failed to build pyaudio
    Running setup.py install for pyaudio ... error
ERROR: Command errored out with exit status 1: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-jq18odkv/pyaudio/setup.py'"'"'; __file__='"'"'/tmp/pip-install-jq18odkv/pyaudio/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-l2sgga5b/install-record.txt --single-version-externally-managed --compile Check the logs for full command output.


In [136]:
# imports
import random
from tqdm import tqdm 
from typing import List
import os, glob, pickle
from pathlib import Path

import librosa
import soundfile

import cv2
import numpy as np
from sklearn.model_selection import train_test_split

from matplotlib import pyplot as plt


import tensorflow as tf
import tensorflow.keras as keras
from tensorflow.keras.models import *
from tensorflow.keras.layers import *
from tensorflow.keras.callbacks import *
from tensorflow.keras.preprocessing.image import ImageDataGenerator

## Audio Preprocessing

In [133]:
# utils

emotions={
    '01':'neutral',
    '02':'calm',
    '03':'happy',
    '04':'sad',
    '05':'angry',
    '06':'fearful',
    '07':'disgust',
    '08':'surprised'
    }

# Emotions to observe
observed_emotions=['calm', 'happy', 'fearful', 'disgust']


def extract_feature(file_name, mfcc, chroma, mel):
  with soundfile.SoundFile(file_name) as sound_file:
      X = sound_file.read(dtype="float32")
      sample_rate=sound_file.samplerate
      if chroma:
          stft=np.abs(librosa.stft(X))
      result=np.array([])
      if mfcc:
        mfccs = np.mean(librosa.feature.mfcc(y=X, sr=sample_rate, n_mfcc=40).T, axis=0)
        mfccs = np.std(mfccs)
        result=np.hstack((result, mfccs))
      if chroma:
        chroma=np.mean(librosa.feature.chroma_stft(S=stft, sr=sample_rate).T,axis=0)
        chroma=np.std(chroma)
        result=np.hstack((result, chroma))
      if mel:
        mel=np.mean(librosa.feature.melspectrogram(X, sr=sample_rate).T,axis=0)
        mel=np.std(mel)
        result=np.hstack((result, mel))
  return result


# load data
def load_data(test_size=0.2):
  x,y=[],[]
  for file in glob.glob(f"{ROOT_SPEECH}/**/*.wav"):
    try:
      file_name=os.path.basename(file)
      emotion=emotions[file_name.split("-")[2]]
      if emotion not in observed_emotions:
        continue
      feature=extract_feature(file, mfcc=True, chroma=True, mel=True)
      x.append(feature)
      y.append(emotion)
    except Exception as e:
      print(f'Unvalidated speech record: {file}')
      continue

  return train_test_split(np.array(x), y, test_size=test_size, random_state=9)
    

In [134]:
# --main
ROOT = './ravdess'
ROOT_SPEECH = './speech'

# unzip Speech Actors
Path(ROOT_SPEECH).mkdir(exist_ok=True)
unzip_cmd = f'unzip {ROOT}/Audio_Speech_Actors_01-24.zip -d ./speech/'
os.system(unzip_cmd)

x_train, x_test, y_train, y_test = load_data(test_size=0.25)
print((x_train.shape[0], x_test.shape[0]))
print(f'Features extracted: {x_train.shape[1]}')

Unvalidated speech record: ./speech/Actor_05/03-01-02-01-02-02-05.wav
Unvalidated speech record: ./speech/Actor_01/03-01-02-01-01-02-01.wav
Unvalidated speech record: ./speech/Actor_20/03-01-06-01-01-02-20.wav
Unvalidated speech record: ./speech/Actor_20/03-01-03-01-02-01-20.wav
(573, 191)
Features extracted: 3


In [135]:
print(x_train[0]) # track data

[1.27670344e+02 4.16242368e-02 1.07234560e-03]


## Video Preprocessing

In [164]:
# utils
def _unzip(path: List[str]):
  unzipped_folders = []

  target = Path(ROOT_DIR, 'videos')
  target.mkdir(exist_ok=True)

  for video_index, video_zip in enumerate(path):  
    source = f'{ROOT_DIR}/{video_zip}'
    print(f'Unzip {source} TO {target}')

    unzip_cmd = f'unzip -q {source} -d {target} &' # umpersant relies for continue of the execution, otherwise STOPS. 
    os.system(unzip_cmd)
    unzipped_folders.append(Path(target, f'Actor_{video_zip.split("_")[3].replace(".zip", "")}'))
  return unzipped_folders

def extract_faces(paths):
  Path('./video_faces/').mkdir(exist_ok=True)
  for folder in unzipped_folders:
    fc = 0
    ret = True
    videos = [file for file in os.listdir(folder)]

    # define detector
    face_cs = cv2.CascadeClassifier()
    face_cs = face_cs.load(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')

    r_frames = [random.randint(0, 20) for _ in range(6)]
    for k in range(len(videos)):
      cap = cv2.VideoCapture(videos[k])
      num, count = 16, 6
      width, height = 64, 64

      buff = np.empty((count, height, width, 3), np.dtype('uint8'))
      all_frames = np.empty((num, height, width, 3), np.dtype('uint8'))
      print(cap)

      while (fc < num and ret):
          ret, original_frame = cap.read()
          original_frame = cv2.cvtColor(original_frame, cv2.COLOR_RGB2BGR)
          if ret:
            #get face
            _frame = cv2.cvtColor(original_frame, cv2.COLOR_BGR2GRAY)
            face = face_cascade.detectMultiScale(_frame)
            if len(face) > 0:
              face = face[0]
              x, y, w, h = [v for v in face]
              cropped_face = original_frame[y:y+h, x:x+w]
              all_frames[fc] = cv2.resize(cropped_face,(64,64))
              fc += 1
                
      for i in range(len(r_frames)):
          buff[i] = all_frames[r_frames[i]]
        
      cap.release()
      f = open("./" + pkl_dir+fnames[k][9:-4] + ".pkl", 'wb')
      pickle.dump(buff, f)
      f.close()
  return True  

In [166]:
# --main
ROOT_DIR = './ravdess'
ROOT_VIDEOS = './ravdess/videos'

video_zips = sorted([zip_name for zip_name in os.listdir(ROOT_DIR) if zip_name.startswith('Video')])  

TOY_TARGET = video_zips[0]

unzipped_folders = _unzip([TOY_TARGET])
extract_faces(unzipped_folders) 

Unzip ./ravdess/Video_Speech_Actor_01.zip TO ravdess/videos


## Models

In [172]:
# Audio Model

# Input
video_input = Input(shape=(24,171,1))

x = Conv2D(32, (3, 3),padding="same", activation='relu', input_shape=(24, 171, 1))(video_input)
x = BatchNormalization()(x)
x = ReLU()(x)
x = MaxPooling2D((2, 2), padding="same")(x)
x = Conv2D(32, (3, 3),padding="same", activation='relu', input_shape=(24, 171, 1))(x)
x = BatchNormalization()(x)
x = ReLU()(x)
x = MaxPooling2D((2, 2), padding="same")(x)
x = Flatten()(x)

# Output
video_output = Dense(128,activation='relu')(x)

video_model = Model(video_input, video_output)
video_model.compile(optimizer='RMSprop', loss='categorical_crossentropy', metrics=['accuracy']) 
video_model.summary()

Model: "functional_16"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_10 (InputLayer)        [(None, 24, 171, 1)]      0         
_________________________________________________________________
conv2d_30 (Conv2D)           (None, 24, 171, 32)       320       
_________________________________________________________________
batch_normalization_10 (Batc (None, 24, 171, 32)       128       
_________________________________________________________________
re_lu_10 (ReLU)              (None, 24, 171, 32)       0         
_________________________________________________________________
max_pooling2d_20 (MaxPooling (None, 12, 86, 32)        0         
_________________________________________________________________
conv2d_31 (Conv2D)           (None, 12, 86, 32)        9248      
_________________________________________________________________
batch_normalization_11 (Batc (None, 12, 86, 32)      

In [173]:
# Audio Model

# Input
audio_input = Input(shape=(6,64,64,3))

x = TimeDistributed(Conv2D(32, (3,3), padding='same', strides=(2,2), activation='relu'))(audio_input)
x = TimeDistributed(Conv2D(32, (3,3), padding='same', strides=(2,2), activation='relu'))(x)
x = TimeDistributed(MaxPooling2D((2,2), strides=(2,2)))(x)
x = TimeDistributed(Conv2D(32, (3,3), padding='same', strides=(2,2), activation='relu'))(x)
x = TimeDistributed(Conv2D(32, (3,3), padding='same', strides=(2,2), activation='relu'))(x)
x = TimeDistributed(MaxPooling2D((2,2), strides=(2,2)))(x)
x = TimeDistributed(GlobalMaxPooling2D())(x)

# Output
audio_output = LSTM(32, activation='relu', return_sequences=False)(x)

audio_model = Model(audio_input, audio_output)
audio_model.compile(optimizer='RMSprop', loss='categorical_crossentropy', metrics=['accuracy']) 
audio_model.summary()

Model: "functional_18"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_11 (InputLayer)        [(None, 6, 64, 64, 3)]    0         
_________________________________________________________________
time_distributed_35 (TimeDis (None, 6, 32, 32, 32)     896       
_________________________________________________________________
time_distributed_36 (TimeDis (None, 6, 16, 16, 32)     9248      
_________________________________________________________________
time_distributed_37 (TimeDis (None, 6, 8, 8, 32)       0         
_________________________________________________________________
time_distributed_38 (TimeDis (None, 6, 4, 4, 32)       9248      
_________________________________________________________________
time_distributed_39 (TimeDis (None, 6, 2, 2, 32)       9248      
_________________________________________________________________
time_distributed_40 (TimeDis (None, 6, 1, 1, 32)     

## Model Fusion

In [183]:
from tensorflow.keras.layers import Concatenate

fusion_concat = Concatenate([audio_model, video_model])
input_size = [audio_input, video_input]

h1 = Dense(256, activation='relu')(fusion_concat)
h2 = Dense(7, activation='softmax')(h1)

model = Model(input_size, h2)
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.summary()